File: UPS.Go.ClientCredentials.SDK/src/ClientCredentialService.go
package clientcredential

import (
    "io"
    "net/http"
    "net/url"
    "strings"
    "time"
    "encoding/json"
    "net"
)

const (
    timeoutDuration       = 30 * time.Second
    tlsHandshakeTimeout   = 10 * time.Second
    idleConnTimeout       = 10 * time.Second
    responseHeaderTimeout = 10 * time.Second
    expectContinueTimeout = 10 * time.Second
    tokenUrl              = "https://onlinetools.ups.com/security/v1/oauth/token"
    timedOut              = `{"response":{"errors":[{"code":"10500","message":"Request Timed out."}]}}`
    internalServerError   = `{"response":{"errors":[{"code":"10500","message":"Internal server error"}]}}`
)

func setHttpClientTimeouts(httpClient *http.Client) *http.Client {
    if httpClient == nil {
        return &http.Client {
            Timeout: timeoutDuration,
            Transport: &http.Transport {
                TLSHandshakeTimeout: tlsHandshakeTimeout,
                IdleConnTimeout: idleConnTimeout,
                ResponseHeaderTimeout: responseHeaderTimeout,
                ExpectContinueTimeout: expectContinueTimeout,
            },
        }
    }

    httpClient.Timeout = timeoutDuration

    if transport, ok := httpClient.Transport.(*http.Transport); ok {
        transport.TLSHandshakeTimeout = tlsHandshakeTimeout
        transport.IdleConnTimeout = idleConnTimeout
        transport.ResponseHeaderTimeout = responseHeaderTimeout
        transport.ExpectContinueTimeout = expectContinueTimeout
    } else {
        httpClient.Transport = &http.Transport {
            TLSHandshakeTimeout: tlsHandshakeTimeout,
            IdleConnTimeout: idleConnTimeout,
            ResponseHeaderTimeout: responseHeaderTimeout,
            ExpectContinueTimeout: expectContinueTimeout,
        }
    }
    return httpClient
}

func GetAccessToken(httpClient *http.Client, clientId string, clientSecret string, headers map[string]string, customClaims map[string]string) (UpsOauthResponse) {
    var hClient *http.Client = setHttpClientTimeouts(httpClient)

    body := url.Values{}
    body.Set("grant_type", "client_credentials")
    body.Set("scope", "public")

    for keys := range customClaims {
        claim := "{\"" + keys + "\":\"" + customClaims[keys] + "\"}"
        body.Add("custom_claims", claim)
    }
    encodedData := body.Encode()

    req,err := http.NewRequest(http.MethodPost, tokenUrl, strings.NewReader(encodedData))
    if err != nil {        
        return apiErrorResponse(internalServerError)
    }

    req.SetBasicAuth(clientId, clientSecret)
    req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

    for keys := range headers {
        req.Header.Set(keys, headers[keys])
    }

    res, err := hClient.Do(req)
    if err != nil {
        if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
            return apiErrorResponse(timedOut)
        }
        return apiErrorResponse(err.Error())
    }

    defer res.Body.Close()

    response, err := io.ReadAll(res.Body)
    if err != nil {
        return apiErrorResponse(internalServerError)
    }

    respString := string(response)
    var data TokenInfo
    errr := json.Unmarshal([]byte(respString), &data)
    if errr != nil {
        return apiErrorResponse(internalServerError)
    } 
    
    if !(res.StatusCode >= 200 && res.StatusCode <= 299) {        
        return apiErrorResponse(respString)
    }

    upsResponse := UpsOauthResponse{
        Response: data,
        Error: "",
    }

    return upsResponse
}

func apiErrorResponse(errorResponse string) UpsOauthResponse {    
    response := UpsOauthResponse{
        Response: TokenInfo{},
        Error: errorResponse,
    }    
    return response
}

type TokenInfo struct {
	Issued_at string `json: "issued_at`
	Token_type string `json: "token_type`
	Client_id string `json: "client_id`
	Access_token string `json: "access_token`
	Expires_in string `json: "expires_in`
	Status string `json: "status`
}

type UpsOauthResponse struct {
	Response TokenInfo
	Error string
}

File: ups-oauth-client-credentials-sdk/src/ClientCredentialService.ts
import { ClientCredentialConstants } from './ClientCredentialConstants';
import { IClientCredentialService, UpsOauthResponse, ErrorResponse } from './IClientCredentialService';
import { TokenInfo } from './TokenInfo';

export class ClientCredentialService implements IClientCredentialService {
    private baseUrl: string;
    private fetchApi: typeof fetch;

    constructor(fetchHttp: typeof fetch) {
        this.baseUrl = ClientCredentialConstants.BASE_URL;
        this.fetchApi = this.validateFetch(fetchHttp);
    }

    public async getAccessToken(clientId, clientSecret, headers = null, customClaims = null): Promise<UpsOauthResponse<TokenInfo>> {
        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), ClientCredentialConstants.PostTimeout);
            const authorization = `Basic ${Buffer.from(`${clientId}:${clientSecret}`).toString('base64')}`;
            const httpHeaders = {
                "Content-Type": "application/x-www-form-urlencoded",
                "Authorization": authorization,
                ...headers
            };

            const body: Record<string, string> = {
                grant_type: 'client_credentials',
                scope: 'public',
                ...(customClaims && { custom_claims: JSON.stringify(customClaims) })
            };

            const customBody = new URLSearchParams(body);
            const response = await this.fetchApi(ClientCredentialConstants.TokenUrl, {
                method: 'POST',
                headers: httpHeaders,
                body: customBody,
                signal: controller.signal
            });

            const json = await response.json();
            clearTimeout(timeoutId);
            if (response.status === 200) {
                return this.apiResponse(json);
            }
            return this.apiErrorResponse(await JSON.stringify(json));
        } catch (error) {
            if (error.name === 'AbortError') {
                const res = JSON.parse(ClientCredentialConstants.TimedOut);
                return this.apiErrorResponse(res);
            } else {
                const res = JSON.parse(ClientCredentialConstants.InternalServerError);
                return this.apiErrorResponse(res);
            }
        }
    }

    private apiResponse(json: any): UpsOauthResponse<TokenInfo> {
        const token: TokenInfo = {
            accessToken: json.access_token,
            clientId: json.client_id,
            expiresIn: json.expires_in,
            issuedAt: json.issued_at,
            status: json.status,
            tokenType: json.token_type
        };
        return {
            response: token,
            error: null
        };
    }

    private apiErrorResponse(json: any): UpsOauthResponse<TokenInfo> {
        const response = json;
        const errorResponse: ErrorResponse = JSON.stringify(response);

        return {
            response: null,
            error: errorResponse
        };
    }

    private validateFetch(fetchHttp: any): typeof fetch {
        if (typeof fetchHttp !== 'function' || !('prototype' in fetchHttp)) {
            return fetch;
        }
        return fetchHttp;
    }
}




